File: agents/go/afl-agent/poller.go
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package aflagent

import (
	"context"
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// Handler is a callback function for processing events.
// It receives the step parameters and returns the result to write back.
type Handler func(params map[string]interface{}) (map[string]interface{}, error)

// AgentPoller polls for tasks and dispatches to registered handlers.
type AgentPoller struct {
	cfg      Config
	serverID string
	db       *mongo.Database
	client   *mongo.Client

	handlers map[string]Handler
	mu       sync.RWMutex

	ops          *MongoOps
	registration *ServerRegistration

	stopCh   chan struct{}
	wg       sync.WaitGroup
	sem      chan struct{} // semaphore for concurrency control
	running  bool
	runMu    sync.Mutex
}

// NewAgentPoller creates a new AgentPoller with the given configuration.
func NewAgentPoller(cfg Config) *AgentPoller {
	return &AgentPoller{
		cfg:      cfg,
		serverID: uuid.New().String(),
		handlers: make(map[string]Handler),
		stopCh:   make(chan struct{}),
		sem:      make(chan struct{}, cfg.MaxConcurrent),
	}
}

// Register registers a handler for a qualified facet name.
// The facet name can be either qualified (ns.FacetName) or short (FacetName).
func (p *AgentPoller) Register(facetName string, handler Handler) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.handlers[facetName] = handler
}

// RegisteredHandlers returns a list of registered handler names.
func (p *AgentPoller) RegisteredHandlers() []string {
	p.mu.RLock()
	defer p.mu.RUnlock()

	names := make([]string, 0, len(p.handlers))
	for name := range p.handlers {
		names = append(names, name)
	}
	return names
}

// Start connects to MongoDB and begins the poll loop.
// This method blocks until Stop is called.
func (p *AgentPoller) Start(ctx context.Context) error {
	p.runMu.Lock()
	if p.running {
		p.runMu.Unlock()
		return nil
	}
	p.running = true
	p.runMu.Unlock()

	// Connect to MongoDB
	clientOpts := options.Client().ApplyURI(p.cfg.MongoURL)
	client, err := mongo.Connect(ctx, clientOpts)
	if err != nil {
		return err
	}
	p.client = client
	p.db = client.Database(p.cfg.Database)
	p.ops = NewMongoOps(p.db)
	p.registration = NewServerRegistration(p.db)

	// Register server
	handlers := p.RegisteredHandlers()
	if err := p.registration.Register(ctx, p.serverID, p.cfg, handlers); err != nil {
		return err
	}

	// Start heartbeat goroutine
	p.wg.Add(1)
	go p.heartbeatLoop(ctx)

	// Run poll loop
	p.pollLoop(ctx)

	return nil
}

// Stop signals the poller to stop and waits for cleanup.
func (p *AgentPoller) Stop(ctx context.Context) error {
	p.runMu.Lock()
	if !p.running {
		p.runMu.Unlock()
		return nil
	}
	p.running = false
	p.runMu.Unlock()

	close(p.stopCh)
	p.wg.Wait()

	// Deregister server
	if p.registration != nil {
		if err := p.registration.Deregister(ctx, p.serverID); err != nil {
			log.Printf("Failed to deregister server: %v", err)
		}
	}

	// Disconnect from MongoDB
	if p.client != nil {
		if err := p.client.Disconnect(ctx); err != nil {
			return err
		}
	}

	return nil
}

// PollOnce performs a single poll cycle. Useful for testing.
func (p *AgentPoller) PollOnce(ctx context.Context) error {
	if p.db == nil {
		// Connect if not already connected
		clientOpts := options.Client().ApplyURI(p.cfg.MongoURL)
		client, err := mongo.Connect(ctx, clientOpts)
		if err != nil {
			return err
		}
		p.client = client
		p.db = client.Database(p.cfg.Database)
		p.ops = NewMongoOps(p.db)
		p.registration = NewServerRegistration(p.db)
	}

	handlers := p.RegisteredHandlers()
	task, err := p.ops.ClaimTask(ctx, handlers, p.cfg.TaskList)
	if err != nil {
		return err
	}
	if task == nil {
		return nil // No task available
	}

	// Process synchronously for PollOnce
	p.processTask(ctx, task)
	return nil
}

func (p *AgentPoller) pollLoop(ctx context.Context) {
	ticker := time.NewTicker(p.cfg.PollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-p.stopCh:
			return
		case <-ctx.Done():
			return
		case <-ticker.C:
			p.pollCycle(ctx)
		}
	}
}

func (p *AgentPoller) pollCycle(ctx context.Context) {
	handlers := p.RegisteredHandlers()
	if len(handlers) == 0 {
		return
	}

	// Try to claim a task
	task, err := p.ops.ClaimTask(ctx, handlers, p.cfg.TaskList)
	if err != nil {
		log.Printf("Error claiming task: %v", err)
		return
	}
	if task == nil {
		return // No task available
	}

	// Acquire semaphore slot
	select {
	case p.sem <- struct{}{}:
		// Got slot, process in goroutine
		p.wg.Add(1)
		go func() {
			defer p.wg.Done()
			defer func() { <-p.sem }()
			p.processTask(ctx, task)
		}()
	default:
		// All slots busy, skip this cycle
		// Task will be picked up next cycle or by another instance
		log.Printf("Max concurrency reached, skipping task %s", task.UUID)
	}
}

// emitStepLog writes a step log entry (best-effort).
func (p *AgentPoller) emitStepLog(ctx context.Context, stepID, workflowID, facetName, level, message string) {
	p.ops.InsertStepLog(ctx, stepID, workflowID, p.serverID, facetName,
		StepLogSourceFramework, level, message)
}

func (p *AgentPoller) processTask(ctx context.Context, task *TaskDocument) {
	// 1. Task claimed
	p.emitStepLog(ctx, task.StepID, task.WorkflowID, task.Name,
		StepLogLevelInfo, fmt.Sprintf("Task claimed: %s", task.Name))

	// Find handler - try qualified name first, then short name
	handler := p.findHandler(task.Name)
	if handler == nil {
		// 2. No handler found
		errMsg := fmt.Sprintf("No handler registered for: %s", task.Name)
		p.emitStepLog(ctx, task.StepID, task.WorkflowID, task.Name,
			StepLogLevelError, "Handler error: "+errMsg)
		log.Printf("No handler for task: %s", task.Name)
		if err := p.ops.MarkTaskFailed(ctx, task, "no handler registered"); err != nil {
			log.Printf("Failed to mark task as failed: %v", err)
		}
		return
	}

	// 3. Dispatching handler
	p.emitStepLog(ctx, task.StepID, task.WorkflowID, task.Name,
		StepLogLevelInfo, fmt.Sprintf("Dispatching handler: %s", task.Name))

	dispatchStart := time.Now()

	// Read step parameters
	params, err := p.ops.ReadStepParams(ctx, task.StepID)
	if err != nil {
		log.Printf("Failed to read step params: %v", err)
		if err := p.ops.MarkTaskFailed(ctx, task, err.Error()); err != nil {
			log.Printf("Failed to mark task as failed: %v", err)
		}
		return
	}

	// Invoke handler
	result, err := handler(params)
	if err != nil {
		// 5. Handler error
		p.emitStepLog(ctx, task.StepID, task.WorkflowID, task.Name,
			StepLogLevelError, fmt.Sprintf("Handler error: %v", err))
		log.Printf("Handler error for %s: %v", task.Name, err)
		if err := p.ops.MarkTaskFailed(ctx, task, err.Error()); err != nil {
			log.Printf("Failed to mark task as failed: %v", err)
		}
		return
	}

	// Write returns to step
	if result != nil {
		if err := p.ops.WriteStepReturns(ctx, task.StepID, result); err != nil {
			log.Printf("Failed to write step returns: %v", err)
			if err := p.ops.MarkTaskFailed(ctx, task, err.Error()); err != nil {
				log.Printf("Failed to mark task as failed: %v", err)
			}
			return
		}
	}

	// Insert resume task for Python RunnerService
	if err := p.ops.InsertResumeTask(ctx, task.StepID, task.WorkflowID, task.TaskListName); err != nil {
		log.Printf("Failed to insert resume task: %v", err)
		if err := p.ops.MarkTaskFailed(ctx, task, err.Error()); err != nil {
			log.Printf("Failed to mark task as failed: %v", err)
		}
		return
	}

	// Mark task completed
	if err := p.ops.MarkTaskCompleted(ctx, task); err != nil {
		log.Printf("Failed to mark task completed: %v", err)
	}

	// 4. Handler completed
	durationMs := time.Since(dispatchStart).Milliseconds()
	p.emitStepLog(ctx, task.StepID, task.WorkflowID, task.Name,
		StepLogLevelSuccess, fmt.Sprintf("Handler completed: %s (%dms)", task.Name, durationMs))
}

func (p *AgentPoller) findHandler(taskName string) Handler {
	p.mu.RLock()
	defer p.mu.RUnlock()

	// Try exact match first
	if h, ok := p.handlers[taskName]; ok {
		return h
	}

	// Try short name fallback (ns.Facet -> Facet)
	if idx := strings.LastIndex(taskName, "."); idx >= 0 {
		shortName := taskName[idx+1:]
		if h, ok := p.handlers[shortName]; ok {
			return h
		}
	}

	return nil
}

func (p *AgentPoller) heartbeatLoop(ctx context.Context) {
	defer p.wg.Done()

	ticker := time.NewTicker(p.cfg.HeartbeatInterval)
	defer ticker.Stop()

	for {
		select {
		case <-p.stopCh:
			return
		case <-ctx.Done():
			return
		case <-ticker.C:
			if err := p.registration.Heartbeat(ctx, p.serverID); err != nil {
				log.Printf("Heartbeat error: %v", err)
			}
		}
	}
}

File: agents/typescript/afl-agent/src/poller.ts
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { MongoClient, Db } from "mongodb";
import { v4 as uuidv4 } from "uuid";
import { AgentPollerConfig } from "./config";
import { MongoOps } from "./mongo-ops";
import { ServerRegistration } from "./server-registration";
import { TaskDocument } from "./models";

/**
 * Handler callback type for processing events.
 */
export type Handler = (
  params: Record<string, unknown>
) => Promise<Record<string, unknown>>;

/**
 * AgentPoller polls for tasks and dispatches to registered handlers.
 */
export class AgentPoller {
  private readonly config: AgentPollerConfig;
  private readonly serverId: string;
  private readonly handlers: Map<string, Handler> = new Map();

  private client: MongoClient | null = null;
  private db: Db | null = null;
  private ops: MongoOps | null = null;
  private registration: ServerRegistration | null = null;

  public metadataProvider: ((facetName: string) => Record<string, unknown> | undefined) | null = null;

  private running = false;
  private pollInterval: NodeJS.Timeout | null = null;
  private heartbeatInterval: NodeJS.Timeout | null = null;
  private activeCount = 0;

  constructor(config: AgentPollerConfig) {
    this.config = config;
    this.serverId = uuidv4();
  }

  /**
   * Registers a handler for a qualified facet name.
   */
  register(facetName: string, handler: Handler): void {
    this.handlers.set(facetName, handler);
  }

  /**
   * Returns a list of registered handler names.
   */
  registeredHandlers(): string[] {
    return Array.from(this.handlers.keys());
  }

  /**
   * Connects to MongoDB and begins the poll loop.
   */
  async start(): Promise<void> {
    if (this.running) {
      return;
    }
    this.running = true;

    // Connect to MongoDB
    this.client = new MongoClient(this.config.mongoUrl);
    await this.client.connect();
    this.db = this.client.db(this.config.database);
    this.ops = new MongoOps(this.db);
    this.registration = new ServerRegistration(this.db);

    // Register server
    const handlers = this.registeredHandlers();
    await this.registration.register(this.serverId, this.config, handlers);

    // Start heartbeat
    this.heartbeatInterval = setInterval(async () => {
      try {
        await this.registration?.heartbeat(this.serverId);
      } catch (err) {
        console.error("Heartbeat error:", err);
      }
    }, this.config.heartbeatIntervalMs);

    // Start poll loop
    this.pollInterval = setInterval(() => {
      this.pollCycle().catch((err) => console.error("Poll cycle error:", err));
    }, this.config.pollIntervalMs);
  }

  /**
   * Stops the poller and cleans up.
   */
  async stop(): Promise<void> {
    if (!this.running) {
      return;
    }
    this.running = false;

    // Stop intervals
    if (this.pollInterval) {
      clearInterval(this.pollInterval);
      this.pollInterval = null;
    }
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
      this.heartbeatInterval = null;
    }

    // Wait for active tasks to complete
    while (this.activeCount > 0) {
      await new Promise((resolve) => setTimeout(resolve, 100));
    }

    // Deregister server
    if (this.registration) {
      try {
        await this.registration.deregister(this.serverId);
      } catch (err) {
        console.error("Failed to deregister server:", err);
      }
    }

    // Disconnect from MongoDB
    if (this.client) {
      await this.client.close();
      this.client = null;
      this.db = null;
      this.ops = null;
      this.registration = null;
    }
  }

  /**
   * Performs a single poll cycle. Useful for testing.
   */
  async pollOnce(): Promise<void> {
    if (!this.db) {
      // Connect if not already connected
      this.client = new MongoClient(this.config.mongoUrl);
      await this.client.connect();
      this.db = this.client.db(this.config.database);
      this.ops = new MongoOps(this.db);
      this.registration = new ServerRegistration(this.db);
    }

    const handlers = this.registeredHandlers();
    if (handlers.length === 0) {
      return;
    }

    const task = await this.ops!.claimTask(handlers, this.config.taskList);
    if (task) {
      await this.processTask(task);
    }
  }

  private async pollCycle(): Promise<void> {
    const handlers = this.registeredHandlers();
    if (handlers.length === 0) {
      return;
    }

    // Check concurrency limit
    if (this.activeCount >= this.config.maxConcurrent) {
      return;
    }

    // Try to claim a task
    const task = await this.ops!.claimTask(handlers, this.config.taskList);
    if (!task) {
      return;
    }

    // Process task asynchronously
    this.activeCount++;
    this.processTask(task)
      .catch((err) => console.error("Task processing error:", err))
      .finally(() => {
        this.activeCount--;
      });
  }

  private async emitStepLog(
    stepId: string,
    workflowId: string,
    facetName: string,
    level: string,
    message: string
  ): Promise<void> {
    try {
      await this.ops!.insertStepLog(
        stepId, workflowId, this.serverId, facetName,
        "framework", level, message
      );
    } catch {
      // best-effort
    }
  }

  private async processTask(task: TaskDocument): Promise<void> {
    // 1. Task claimed
    await this.emitStepLog(task.step_id, task.workflow_id, task.name,
      "info", `Task claimed: ${task.name}`);

    // Find handler
    const handler = this.findHandler(task.name);
    if (!handler) {
      // 2. No handler found
      await this.emitStepLog(task.step_id, task.workflow_id, task.name,
        "error", `Handler error: No handler registered for: ${task.name}`);
      console.log(`No handler for task: ${task.name}`);
      await this.ops!.markTaskFailed(task, "no handler registered");
      return;
    }

    try {
      // 3. Dispatching handler
      await this.emitStepLog(task.step_id, task.workflow_id, task.name,
        "info", `Dispatching handler: ${task.name}`);

      const dispatchStart = Date.now();

      // Read step parameters
      const params = await this.ops!.readStepParams(task.step_id);

      // Inject handler-level step_log callback
      params["_step_log"] = async (message: string, level = "info") => {
        try {
          await this.ops!.insertStepLog(task.step_id, task.workflow_id,
            this.serverId, task.name, "handler", level, message);
        } catch { /* best-effort */ }
      };

      // Inject _facet_name
      params["_facet_name"] = task.name;

      // Inject _update_step callback for streaming partial results
      params["_update_step"] = async (partial: Record<string, unknown>) => {
        try {
          await this.ops!.updateStepReturns(task.step_id, partial);
        } catch { /* best-effort */ }
      };

      // Inject _handler_metadata if provider is available
      if (this.metadataProvider) {
        const meta = this.metadataProvider(task.name);
        if (meta) {
          params["_handler_metadata"] = meta;
        }
      }

      // Invoke handler
      const result = await handler(params);

      const durationMs = Date.now() - dispatchStart;

      // Write returns to step
      if (result && Object.keys(result).length > 0) {
        await this.ops!.writeStepReturns(task.step_id, result);
      }

      // Insert resume task for Python RunnerService
      await this.ops!.insertResumeTask(
        task.step_id,
        task.workflow_id,
        task.task_list_name,
        task.name
      );

      // Mark task completed
      await this.ops!.markTaskCompleted(task);

      // 4. Handler completed
      await this.emitStepLog(task.step_id, task.workflow_id, task.name,
        "success", `Handler completed: ${task.name} (${durationMs}ms)`);
    } catch (err) {
      // 5. Handler error
      const errorMsg = err instanceof Error ? err.message : String(err);
      await this.emitStepLog(task.step_id, task.workflow_id, task.name,
        "error", `Handler error: ${errorMsg}`);
      console.error(`Handler error for ${task.name}:`, err);
      await this.ops!.markTaskFailed(task, errorMsg);
    }
  }

  private findHandler(taskName: string): Handler | undefined {
    // Try exact match first
    const exact = this.handlers.get(taskName);
    if (exact) {
      return exact;
    }

    // Try short name fallback (ns.Facet -> Facet)
    const lastDot = taskName.lastIndexOf(".");
    if (lastDot >= 0) {
      const shortName = taskName.substring(lastDot + 1);
      return this.handlers.get(shortName);
    }

    return undefined;
  }
}
